internal/storage/blob/cloner.go

```diff
@@ -91,45 +76,83 @@ func (c *Cloner) Clone(ctx context.Context) (*CloneResult, error) {
 			c.log.Errorw("Failed to get next item", "error", err)
 			return nil, fmt.Errorf("failed to get next object in the bucket: %w", err)
 		}
+
 		file := strings.TrimPrefix(obj.Key, "/")
-		eTag := obj.MD5
 		if util.FileType(file) == util.FileTypeNotIndexed {
 			continue
 		}
-		info[file] = eTag
-		if eTag != nil && bytes.Equal(eTag, c.info[file]) {
+
+		if obj.MD5 == nil {
+			return nil, fmt.Errorf("blob object doesn't have md5 specified: %w", err)
+		}
+
+		if !util.IsSupportedFileType(file) {
+			c.log.Debugw("Unsupported file is ignored", "file", file)
+			continue
+		}
+
+		etag := fmt.Sprintf("%x", obj.MD5)
+		all[etag] = append(all[etag], file)
+
+		if existingFiles, ok := c.state[etag]; ok {
+			if !slices.Contains(existingFiles, file) {
+				addedOrUpdated = append(addedOrUpdated, info{
+					etag: etag,
+					file: file,
+				})
+			}
+
 			continue
 		}
-		if err = c.downloadToFile(ctx, obj.Key, file); err != nil {
-			c.log.Errorw("Failed to download file", "error", err, "file", file)
-			cr.failuresCount++
+
+		if err = c.downloadToFile(ctx, obj.Key, etag); err != nil {
+			c.log.Errorw("Failed to download file", "error", err, "etag", etag, "file", file)
+			failuresCount++
 		} else {
-			cr.updateOrAdd = append(cr.updateOrAdd, fileInfo{file: file, etag: eTag})
+			addedOrUpdated = append(addedOrUpdated, info{
+				etag: etag,
+				file: file,
+			})
 		}
 	}
 
-	for key := range c.info {
-		if _, ok := info[key]; !ok {
-			c.log.Debugw("Removing file", "file", key)
-			err := c.fsys.Remove(key)
-			if err != nil {
-				return nil, err
+	var danglingEtags []string
+	var deleted []info
+	for etag, existingFiles := range c.state {
+		for _, existingFile := range existingFiles {
+			if files, ok := all[etag]; !ok {
+				deleted = append(deleted, info{
+					etag: etag,
+					file: existingFile,
+				})
+
+				// This etag is not referenced from any file, we should get rid of it later.
+				danglingEtags = append(danglingEtags, etag)
+			} else if !slices.Contains(files, existingFile) {
+				deleted = append(deleted, info{
+					etag: etag,
+					file: existingFile,
+				})
 			}
-			cr.delete = append(cr.delete, key)
 		}
 	}
-	c.info = info
-	return cr, nil
+
+	c.danglingEtags = danglingEtags
+	c.state = all
+	return &CloneResult{
+		all:            all,
+		addedOrUpdated: addedOrUpdated,
+		deleted:        deleted,
+		failuresCount:  failuresCount,
+	}, nil
 }
 
-func (c *Cloner) downloadToFile(ctx context.Context, key, file string) (err error) {
-	// Create the directories in the path
+func (c *Cloner) downloadToFile(ctx context.Context, key, file string) error {
```

Comment:
```suggestion
func (c *Cloner) downloadToFile(ctx context.Context, key, file string) (err error) {
```

Removing the output variable makes the `multierr.AppendInvoke` on line 160 not have an effect.